### PR TITLE
Make link fixing for Robot functional

### LINKED ISSUE #83: Get rid of imperative "fixing"
...of links.
Frank

### PR DESCRIPTION
**Pull request by varunagrawal**
Fixes #83

### CODE REVIEW

File: gtdynamics/universal_robot/Robot.cpp [hunk 1]
```diff
@@ -82,8 +82,10 @@ Robot Robot::fixLink(const std::string &name) {
   if (name_to_link_.find(name) == name_to_link_.end()) {
     throw std::runtime_error("no link named " + name);
   }
-  name_to_link_.at(name)->fix();
-  return *this;
+
+  Robot fixed_robot = Robot(*this);
+  fixed_robot.name_to_link_.at(name)->fix();
```
Comment (dellaert) [inline gtdynamics/universal_robot/Robot.cpp:87]: so, is fix now private? Otherwise this will become an issue again :-)

**Review by dellaert — APPROVED** [review]

File: gtdynamics/universal_robot/Robot.cpp (hunk 1 above)
Comment (varunagrawal) [reply gtdynamics/universal_robot/Robot.cpp:87]: Ah good catch.

Comment (varunagrawal) [reply gtdynamics/universal_robot/Robot.cpp:87]: Okay so `sdf.cpp` needs the `fix` method, so I'm making both `Link::fix` and `Link::unfix` functional as well.

Comment (varunagrawal) [reply gtdynamics/universal_robot/Robot.cpp:87]: I figured out a cleaner way.

**Comment by varunagrawal** [comment]
Re-requesting review since there are some substantial changes.

**Comment by varunagrawal** [comment]
@dellaert gentle reminder so I can merge this in before tomorrow's meeting.

File: gtdynamics.i [hunk 2]
```diff
@@ -200,6 +197,11 @@ class Link  {
   gtsam::Matrix6 inertiaMatrix();
 
   void print(const std::string &s = "") const;
+
+  static gtdynamics::Link fix(const gtdynamics::Link& link);
```
Comment (dellaert) [inline gtdynamics.i:201]: Why expose them in first place if they are private?

File: gtdynamics/universal_robot/Joint.h [hunk 3]
```diff
@@ -75,6 +75,10 @@ struct JointParams {
 
 /// Joint is the base class for a joint connecting two Link objects.
 class Joint : public boost::enable_shared_from_this<Joint> {
+
+  /// Robot class should have access to the internals of its joints.
```
Comment (dellaert) [inline gtdynamics/universal_robot/Joint.h:79]: Why?

**Review by dellaert — APPROVED** [review]

File: gtdynamics.i (hunk 2 above)
Comment (varunagrawal) [reply gtdynamics.i:201]: This is the static (functional) version.

File: gtdynamics/universal_robot/Joint.h (hunk 3 above)
Comment (varunagrawal) [reply gtdynamics/universal_robot/Joint.h:79]: Logically makes sense right? A robot should be able to inspect its joints and links.

**Pull request merged**